src/TransformerGo.cs

```diff
@@ -75,54 +67,48 @@ private void TransformEnumTypes(CodeModelGo cmg)
             }
 
             // Add discriminators
-            foreach (var mt in cmg.ModelTypes)
+            foreach (var mt in cmg.ModelTypes.Cast<CompositeTypeGo>())
             {
-                if (mt.IsPolymorphic)
+                if (!mt.IsPolymorphic)
+                {
+                    continue;
+                }
+
+                var enumValues = new List<EnumValue>();
+
+                var baseTypeEnumValue = new EnumValue
                 {
-                    var values = new List<EnumValue>();
+                    Name = $"{CodeNamerGo.Instance.GetTypeName(mt.PolymorphicDiscriminator)}{CodeNamerGo.Instance.GetTypeName(mt.SerializedName)}",
+                    SerializedName = mt.SerializedName
+                };
 
-                    var baseTypeEnumValue = new EnumValue
+                enumValues.Add(baseTypeEnumValue);
+
+                foreach (var dt in mt.DerivedTypes)
+                {
+                    var ev = new EnumValue
                     {
-                        Name = $"{CodeNamerGo.Instance.GetTypeName(mt.PolymorphicDiscriminator)}{CodeNamerGo.Instance.GetTypeName(mt.SerializedName)}",
-                        SerializedName = mt.SerializedName
+                        Name = $"{CodeNamerGo.Instance.GetTypeName(mt.PolymorphicDiscriminator)}{CodeNamerGo.Instance.GetTypeName(dt.SerializedName)}",
+                        SerializedName = dt.SerializedName
                     };
+                    enumValues.Add(ev);
+                }
+                bool nameAlreadyExists = cmg.EnumTypes.Any(et => et.Name.EqualsIgnoreCase(mt.PolymorphicDiscriminator));
+                bool enumAlreadyExists = nameAlreadyExists;
 
-                    values.Add(baseTypeEnumValue);
+                if (nameAlreadyExists)
+                {
+                    mt.DiscriminatorEnum = (EnumTypeGo) cmg.EnumTypes.First(et => et.Name.EqualsIgnoreCase(mt.PolymorphicDiscriminator));
+                    enumAlreadyExists = !mt.DiscriminatorEnum.Values.Select(value => value.SerializedName)
+                        .Except(enumValues.Select(value => value.SerializedName)).Any();

```

Comment:
is the intent here that all the discriminator values are part of the existing enum type or that they are equal ? if it is the later we can use `SequenceEqual` instead of `Except` . I used `Except` to keep the same behavior as previously

Reply:
I believe it's the former but not 100% sure, maybe @mcardosos can comment further as she did the original implementation.

Reply:
@ 

---
In reply to: [155364263](https://github.com/Azure/autorest.go/pull/60#discussion_r155364263) [](ancestors = 155364263)

Reply:
They all should be part of the already exisitng enum. If not, that is when a new enum should be created.

---
In reply to: [155364263](https://github.com/Azure/autorest.go/pull/60#discussion_r155364263) [](ancestors = 155364263)